sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/subquery.scala

```diff
@@ -316,25 +316,46 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
     newExpression.asInstanceOf[E]
   }
 
+  private def removeAlias(expr: Expression): Expression = expr match {
+    case Alias(c, _) => removeAlias(c)
+    case _ => expr
+  }
+
   /**
    * Statically evaluate an expression containing zero or more placeholders, given a set
-   * of bindings for placeholder values.
+   * of bindings for placeholder values, if the expression is evaluable. If it is not,
+   * bind statically evaluated expression results to an expression.
    */
-  private def evalExpr(expr: Expression, bindings: Map[ExprId, Option[Any]]) : Option[Any] = {
+  private def bindingExpr(
+      expr: Expression,
+      bindings: Map[ExprId, Option[Expression]]): Option[Expression] = {
     val rewrittenExpr = expr transform {
       case r: AttributeReference =>
         bindings(r.exprId) match {
-          case Some(v) => Literal.create(v, r.dataType)
+          case Some(v) => v
           case None => Literal.default(NullType)
         }
     }
-    Option(rewrittenExpr.eval())
+
+    // Removes Alias over given expression, because Alias is not foldable.
+    if (!removeAlias(rewrittenExpr).foldable) {
+      // SPARK-28441: Some expressions, like PythonUDF, can't be statically evaluated.
+      // Needs to evaluate them on query runtime.
+      Some(rewrittenExpr)
+    } else {
+      val exprVal = rewrittenExpr.eval()
+      if (exprVal == null) {
+        None
```

Comment:
Do you know why we need to return `None` here instead of a null literal?

Reply:
I think it uses None to make checking bindings easier.

In other way, to use null literal, `Option[Expression]` can be changed to `Expression` in methods like `evalSubqueryOnZeroTups`, `evalPlan`. Then we check bindings by literal instead of None. Good thing is we can write `Literal.create(rewrittenExpr.eval(), expr.dataType)`, instead of checking null. Looks like just a choice problem.